src/main/kotlin/org/jitsi/nlj/rtcp/KeyframeRequester.kt

```diff
@@ -203,6 +224,26 @@ class KeyframeRequester : TransformerNode("Keyframe Requester") {
                     localSsrc = event.ssrc
                 }
             }
+            is ReceiveSsrcAddedEvent -> {
+                if (event.mediaType == MediaType.VIDEO) {
+                    receiveVideoSsrcs.add(event.ssrc)
+                }
+            }
+            is ReceiveSsrcRemovedEvent -> {
+                receiveVideoSsrcs.remove(event.ssrc)
+                if (event.ssrc == keyframeSsrc) {
+                    keyframeSsrc = null
+                }
+            }
+            is LocalSsrcAssociationEvent -> {
+                // NOTE(brian): this relies on getting the receive SSRCs before we get the
```

Comment:
I worry that if something changes (or in rare scenarios, e.g. someone joining without video and enabling it later) and we hit this bug, it will be hard to find. It shouldn't be hard to fix here by storing the associations, but I leave it to you to decide if you want to do it.

I wonder if we can improve the code which handles these events. It's nice that `Node`s keep their own state because they don't have to reach into a central place, but we now have a bunch of duplicated code because of this and it keeps getting more complex. I think we should move this logic to a separate class to hold SSRCs, SSRC associations, payload types and RTCP features. `Node`s will then forward the events to their own instance. We could also make the class immutable and keep a single copy-on-write instance in the `Transceiver`, which would allows us to remove most of the event handling code.

Reply:
yeah i think this is probably a good idea (at least the helper class and maybe holding it in the transceiver).  if you agree, i think i'll do this in a follow up PR (which i'll work on next).